### PR TITLE
[IMP] project: add personal stages

### PR DESCRIPTION
**Pull request by Williambraecky**
Add user personal stages on tasks, the goal is to allow any user to have their own 'pipeline', for each task they are assigned to they will be able to assign a personal stage which will be different from the one used in the project pipeline.

Task ID: 2398734

### CODE REVIEW

File: addons/project/views/project_views.xml [hunk 1]
```diff
@@ -708,7 +709,7 @@
                     <field name="allow_task_dependencies" invisible="1" />
                     <header>
                         <button name="action_assign_to_me" string="Assign to Me" type="object" class="oe_highlight"
-                            attrs="{'invisible' : [('user_id', '!=', False)]}" data-hotkey="q"/>
+                            attrs="{'invisible' : [('user_ids', '!=', [])]}" data-hotkey="q"/>
```
Comment (rco-odoo) [inline addons/project/views/project_views.xml:712]:
```suggestion
                            attrs="{'invisible' : [('user_ids', '!=', False)]}" data-hotkey="q"/>
```

File: odoo/models.py [hunk 2]
```diff
@@ -2334,7 +2334,7 @@ def _read_group_format_result(self, data, annotated_groupbys, groupby, domain):
             # full domain for this groupby spec
             d = None
             if value:
-                if ftype == 'many2one':
+                if ftype in ['many2one', 'many2many']:
```
Comment (rco-odoo) [inline odoo/models.py:2337]: tuples are faster (VM optimization)
```suggestion
                if ftype in ('many2one', 'many2many'):
```

File: odoo/models.py [hunk 3]
```diff
@@ -2597,12 +2597,12 @@ def _read_group_raw(self, domain, fields, groupby, offset=0, limit=None, orderby
             )
         return result
 
-    def _read_group_resolve_many2one_fields(self, data, fields):
-        many2onefields = {field['field'] for field in fields if field['type'] == 'many2one'}
-        for field in many2onefields:
+    def _read_group_resolve_many2x_fields(self, data, fields):
+        many2xfields = {field['field'] for field in fields if field['type'] in ['many2one', 'many2many']}
```
Comment (rco-odoo) [inline odoo/models.py:2601]:
```suggestion
        many2xfields = {field['field'] for field in fields if field['type'] in ('many2one', 'many2many')}
```

**Review by rco-odoo — COMMENTED** [review]
You really need to write some tests with all possible assignments (no assigned user, single assigned user, multiple assigned user).

**Comment by rco-odoo** [comment]
The change in the result of `read_group` with many2many fields is fine by me.
It totally makes sense, since the value is returned like a many2one.

File: addons/project/models/project.py [hunk 4]
```diff
@@ -858,6 +868,41 @@ def _read_group_stage_ids(self, stages, domain, order):
...
+        for personal_stage in personal_stages:
+            personal_stage.task_id.personal_stage_id = personal_stage
+        # Make sure we assign false to all other tasks
+        for task in self:
+            if not task.personal_stage_id:
+                task.personal_stage_id = False
+
+    def _populate_missing_personal_stages(self):
+        # Assign the default personal stage when the users for those that are missing
+        personal_stages_without_stage = self.env['project.task.stage.personal'].search([('task_id', 'in', self.ids), ('stage_id', '=', False)])
```
[the remark this answers is not in the document]
Comment (Williambraecky) [reply addons/project/models/project.py:891]: I don't think this one needs to check for `user_id` as it is only used to populate those that are missing a `stage_id`

File: addons/project/models/project.py [hunk 5]
```diff
@@ -858,6 +868,41 @@ def _read_group_stage_ids(self, stages, domain, order):
...
+            personal_stages_without_stage.write({'stage_id': stage.id})
+
+    @api.depends('personal_stage_id')
+    def _compute_personal_stage_type_id(self):
+        for task in self:
+            task.personal_stage_type_id = task.personal_stage_id.stage_id
+
+    def _inverse_personal_stage_type_id(self):
+        for task in self:
+            task.personal_stage_id.stage_id = task.personal_stage_type_id
```
[the remark this answers is not in the document]
Comment (Williambraecky) [reply addons/project/models/project.py:904]: There is a reason (which I should have documented ... ), I can not remember right now, but I will test again with a related field

File: addons/project/models/project.py [hunk 6]
```diff
@@ -1248,7 +1294,47 @@ def write(self, vals):
...
+        if x2m_tracking_values:
+            for task, tracking_values in x2m_tracking_values.items():
+                # Compile the different changes
+                MailTracking = self.env['mail.tracking.value']
+                tracking_value_ids = []
+                # Use mail.tracking.value to track our changes, this is to use the same format as the default tracking one
+                #  we just hack the record to think he is a text field and compile the data ourself beforehand
+                for field in tracking_values:
+                    if task[field] != tracking_values[field][0]:
+                        field_desc = task._fields[field].string
```
[the remark this answers is not in the document]
Comment (Williambraecky) [reply addons/project/models/project.py:1327]: How may I have the translated name?

File: addons/project/models/project.py [hunk 7]
```diff
@@ -1248,7 +1294,47 @@ def write(self, vals):
...
+        #         field_name: (value, display_value)
+        #     }
+        # }
+        if x2m_vals_common_fields:
+            # Compute the value before the changes
+            for task in self:
+                task_values = x2m_tracking_values.setdefault(task, {})
+                for field in x2m_vals_common_fields:
+                    if getattr(task, field, None) is None or not isinstance(task[field], Iterable):
+                        continue
```
[the remark this answers is not in the document]
Comment (Williambraecky) [reply addons/project/models/project.py:1314]: Is is indeed redundant to check for this as long as the input is correct

File: addons/project/models/project.py [hunk 8]
```diff
@@ -1248,7 +1294,47 @@ def write(self, vals):
...
+                    task_values[field] = (task[field], ', '.join(record.display_name for record in task[field]))
```
[the remark this answers is not in the document]
Comment (Williambraecky) [reply addons/project/models/project.py:1315]: Would you recommend tracking just `name`?

File: addons/project/models/project.py [hunk 9]
```diff
@@ -1270,14 +1356,67 @@ def _unlink_except_recurring(self):
...
+
     # ---------------------------------------------------
     # Subtasks
     # ---------------------------------------------------
 
     @api.depends('parent_id.user_id')
     def _compute_user_id(self):
         for task in self:
-            task.user_id = task.user_id or task.parent_id.user_id or self.env.uid
+            task.user_ids |= task.parent_id.user_ids or self.env.uid
```
[the remark this answers is not in the document]
Comment (Williambraecky) [reply addons/project/models/project.py:1419]: Indeed

**Comment by Williambraecky** [comment]
> You really need to write some tests with all possible assignments (no assigned user, single assigned user, multiple assigned user).

Many thanks for the review! I've added small comments on things I am not quite sure about right now..

**Review by tivisse — COMMENTED** [review]
Translation comment.

File: addons/project/models/project.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (tivisse) [reply addons/project/models/project.py:1616]: You can use ``def _description_string``

[the remark this answers is not in the document]
Comment (Williambraecky) [reply addons/project/models/project.py:1616]: Does indeed work, thanks

**Comment by tivisse** [comment]
@robodoo r+

**Comment by tivisse** [comment]
@robodoo r+

**Comment by tivisse** [comment]
@robodoo r+

**Comment by tivisse** [comment]
@robodoo r+ rebase-ff

File: addons/project/models/project_task_stage_personal.py [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+    _description = 'Personal Task Stage'
+    _table = 'project_task_user_rel'
+    _rec_name = 'stage_id'
+
+    task_id = fields.Many2one('project.task', required=True, ondelete='cascade', index=True)
+    user_id = fields.Many2one('res.users', required=True, ondelete='cascade', index=True)
+    stage_id = fields.Many2one('project.task.type', domain="[('user_id', '=', user_id)]", ondelete='restrict')
+
+    _sql_constraints = [
+        ('project_personal_stage_unique', 'UNIQUE (task_id, user_id)', 'A task can only have a single personal stage by user.'),
```
Comment (KangOl) [inline addons/project/models/project_task_stage_personal.py:17]:
```suggestion
        ('project_personal_stage_unique', 'UNIQUE (task_id, user_id)', 'A task can only have a single personal stage per user.'),
```

**Comment by tivisse** [comment]
@robodoo r+

**Comment by tivisse** [comment]
@robodoo r+

**Pull request closed without merging**